compiler/rustc_infer/src/infer/resolve.rs

```diff
@@ -16,26 +16,27 @@ use std::ops::ControlFlow;
 /// useful for printing messages etc but also required at various
 /// points for correctness.
 pub struct OpportunisticVarResolver<'a, 'tcx> {
-    infcx: &'a InferCtxt<'tcx>,
+    shallow_resolver: crate::infer::ShallowResolver<'a, 'tcx>,
```

Comment:
```suggestion
    // We use a shallow resolver recursively to resolve inference variables at every level of the type.
    shallow_resolver: crate::infer::ShallowResolver<'a, 'tcx>,
```

Or if you think that's dumb, maybe just a comment saying that we're inlining `infcx.shallow_resolve` :smiley: 

This could use a comment just to explain why we're using shallow resolver in a deep resolver, perhaps